src/matter/cluster/OnOffCluster.ts

```diff
@@ -4,18 +4,109 @@
+}
+const OnOffDyingLightEffectVariantT = { tlvType: TlvType.UnsignedInt } as Template<OnOffDyingLightEffectVariant>;
+
+/*
+  <bitmap name="OnOffControl" type="BITMAP8">
+    <cluster code="0x0006"/>
+    <field name="AcceptOnlyWhenOn" mask="0x01"/>
+  </bitmap>
+*/
+
+const OffWithEffectRequestT = ObjectT({
+    effectId: Field(0, OnOffEffectIdentifierT),
+    effectVariant: Field(1, OnOffDelayedAllOffEffectVariantT), /* OR OnOffDyingLightEffectVariantT depending on first field */
+});
+
+const OnWithTimedOffRequestT = ObjectT({
+    onOffControl: Field(0, UnsignedIntT), /* TODO: Type is a Bit map! */
+    onTime: Field(1, UnsignedIntT), /* min:0, max: 254 */
+    offWaitTime: Field(2, UnsignedIntT), /* min:0, max: 254 */
+});
+
+/**
+ * Attributes and commands for switching devices between 'On' and 'Off' states.
+ */
 export const OnOffCluster = Cluster(
     0x06,
     "On/Off",
     {
-        on: Attribute(0, BooleanT),
+        onOff: Attribute(0, BooleanT, false), /* reportable: true */
+        globalSceneControl: OptionalAttribute(0x4000, BooleanT, true),
+        onTime: OptionalWritableAttribute(0x4001, UnsignedIntT, 0),
+        offWaitTime: OptionalWritableAttribute(0x4002, UnsignedIntT, 0),
+        startUpOnOff: OptionalWritableAttribute(0x4003, OnOffStartUpOnOffT), /* nullable: true, writeAcl: manage */
     },
     {
+        /**
+         * On receipt of this command, a device SHALL enter its ‘Off’ state. This state is device dependent, but it is
+         * recommended that it is used for power off or similar functions. On receipt of the Off command, the OnTime
+         * attribute SHALL be set to 0.
+         */
         off: Command(0, NoArgumentsT, 0, NoResponseT),
+        /**
+         * On receipt of this command, a device SHALL enter its ‘On’ state. This state is device dependent, but it is
+         * recommended that it is used for power on or similar functions. On receipt of the On command, if the value
+         * of the OnTime attribute is equal to 0, the device SHALL set the OffWaitTime attribute to 0.
+         */
         on: Command(1, NoArgumentsT, 1, NoResponseT),
+        /**
+         * On receipt of this command, if a device is in its ‘Off’ state it SHALL enter its ‘On’ state. Otherwise,
+         * if it is in its ‘On’ state it SHALL enter its ‘Off’ state. On receipt of the Toggle command, if the value
+         * of the OnOff attribute is equal to FALSE and if the value of the OnTime attribute is equal to 0, the device
+         * SHALL set the OffWaitTime attribute to 0. If the value of the OnOff attribute is equal to TRUE, the OnTime
+         * attribute SHALL be set to 0.
+         */
         toggle: Command(2, NoArgumentsT, 2, NoResponseT),
+        /**
+         * The OffWithEffect command allows devices to be turned off using enhanced ways of fading.
+         */
+        offWithEffect: Command(0x40, OffWithEffectRequestT, 0x40, NoResponseT), /* optional: true */
```

Comment:
Should we add an "OptionalCommand" which is then also optional for implementing the cluster logic?

Reply:
added TODOs for it